### PR TITLE
Fallback bootstrap router build parameter, ignore obsolete bootstrap routers

### PR DESCRIPTION
**Pull request by tewinget**
Addresses #1995 .

### CODE REVIEW

File: llarp/router_contact.cpp [hunk 1]
```diff
@@ -532,6 +532,21 @@ namespace llarp
     return false;
   }
 
+  bool
+  RouterContact::IsObsoleteBootstrap() const
+  {
+    constexpr std::array obsolete_bootstraps = {
```
Comment (jagerman) [inline llarp/router_contact.cpp]: Should move this outside of the function (and make it static constexpr)

File: llarp/router/router.cpp [hunk 2]
```diff
@@ -666,39 +666,24 @@ namespace llarp
...
+    // that no longer exists
+    for (auto rc_itr = b_list.begin(); rc_itr != b_list.end();)
+    {
+      if (rc_itr->IsObsoleteBootstrap())
+        b_list.erase(rc_itr);
+      else
+        rc_itr++;
+    }
+
     for (auto& rc : b_list)
```
Comment (jagerman) [inline llarp/router/router.cpp]: This for loop can be DRYed out a bit with the nearly identical loop below

File: llarp/router/router.cpp [hunk 3]
```diff
@@ -711,6 +696,24 @@ namespace llarp
 
     if (bootstrapRCList.empty() and not conf.bootstrap.seednode)
     {
+#ifdef BOOTSTRAP_FALLBACK
```
Comment (jagerman) [inline llarp/router/router.cpp]: I would rather keep the `ifdef` as small as possible to minimize code compilation differences, perhaps via this outside the function:
```C++
static constexpr std::optional<std::string_view> bootstrap_fallback_path{
#ifdef BOOTSTRAP_FALLBACK
    std::string_view{BOOTSTRAP_FALLBACK}
#else
    std::nullopt;
#endif
};
```
Then the code here can do (without an ifdef):
```C++
if (bootstrap_fallback_path)
{
  ...
}
```

File: llarp/router/router.cpp [hunk 4]
```diff
@@ -666,52 +666,57 @@ namespace llarp
...
 
     if (bootstrapRCList.empty() and not conf.bootstrap.seednode)
     {
-      throw std::runtime_error{"we have no bootstrap nodes"};
+      if (not bootstrap_fallback.empty())
+      {
+        b_list.clear();
+        b_list.AddFromFile(BOOTSTRAP_FALLBACK);
+
+        verifyRCs(false);
```
Comment (jagerman) [inline llarp/router/router.cpp]:
```suggestion
        verifyRCs(true);
```

File: llarp/router/router.cpp [hunk 5]
```diff
@@ -666,52 +666,57 @@ namespace llarp
...
+      for (auto& rc : b_list)
       {
-        LogWarn("ignoring invalid RC: ", RouterID(rc.pubkey));
-        continue;
+        if (not rc.Verify(Now()))
+        {
+          LogWarn("ignoring invalid RC: ", RouterID(rc.pubkey));
+          continue;
+        }
+        if (ignoreObsolete and rc.IsObsoleteBootstrap())
```
Comment (jagerman) [inline llarp/router/router.cpp]: Should ~~LogWarn~~ `log::warning(logcat, ...)` here as well to let the admin know there is something being skipped in the bootstrap file.

File: llarp/router/router.cpp [hunk 6]
```diff
@@ -1032,14 +1037,27 @@ namespace llarp
...
+        m_NextDecommissionWarn = now + DecommissionWarnInterval;
+      }
+      else if (isSvcNode)
+      {
+        constexpr int KnownPeerWarningThreshold = 5;
+        if (nodedb()->NumLoaded() < KnownPeerWarningThreshold)
+          LogError(
+              "We appear to be an active service node, but have fewer than ",
+              KnownPeerWarningThreshold,
+              " known peers.");
```
Comment (jagerman) [inline llarp/router/router.cpp]:
```suggestion
          log::error(logcat,
              "We appear to be an active service node, but have fewer than {} known peers", KnownPeerWarningThreshold);
```

File: llarp/router/router.cpp [hunk 7]
```diff
@@ -666,52 +666,60 @@ namespace llarp
...
+    }
+
+    auto verifyRCs = [&]() {
+      for (auto& rc : b_list)
       {
-        LogWarn("ignoring invalid RC: ", RouterID(rc.pubkey));
-        continue;
+        if (rc.IsObsoleteBootstrap())
+        {
+          LogWarn("ignoring obsolete boostrap RC: ", RouterID(rc.pubkey));
```
Comment (jagerman) [inline llarp/router/router.cpp]:
```suggestion
          log::warning(logcat, "ignoring obsolete boostrap RC: {}", RouterID(rc.pubkey));
```

File: llarp/bootstrap.cpp [hunk 8]
```diff
@@ -36,4 +36,35 @@ namespace llarp
...
+    }
+    else
+    {
+      RouterContact rc;
+      if (not rc.Read(fpath))
+      {
+        throw std::runtime_error{
+            fmt::format("failed to decode bootstrap RC, file='{}', rc={}", fpath, rc)};
+      }
+      this->insert(rc);
```
Comment (majestrate) [inline llarp/bootstrap.cpp:67]: since you're here, we really need to not make BootstrapList a subtype of std::vector

Comment (jagerman) [reply llarp/bootstrap.cpp:67]: ... the fuck?

Comment (majestrate) [reply llarp/bootstrap.cpp:67]: my bad. std::set

Comment (jagerman) [reply llarp/bootstrap.cpp:67]: Should be done, but seems a bit out of scope for this PR.

Comment (tewinget) [reply llarp/bootstrap.cpp:67]: std::set, and while that is something that would be good it's out of scope for this PR

**Review by majestrate — APPROVED** [review]

**Comment by jagerman** [comment]
One more thing:

Let's make lokinet-router *not* ping oxend if it's active and doesn't know of at least (threshold) number of routers, because that generally indicates some sort of serious connectivity issue and we want the operator to notice right away (because of a lack of proofs).  (We *do* still want it to ping oxend while unfunded, though, because it's common for an unfunded router to not know about any routers).

**Review by jagerman — CHANGES_REQUESTED** [review]
Two changes needed here:

- the "fallback" should just bake the bootstrap data into the binary at compile time rather than carrying a path around that we have to distribute.
- we need to support bootstraps for different networks, typically mainnet + testnet, so that compile time fallback should probably be a unordered_map of netid -> bootstrap data.

**Review by jagerman — APPROVED** [review]

**Pull request merged**